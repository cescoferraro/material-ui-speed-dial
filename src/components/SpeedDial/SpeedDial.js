// @flow

import React, { Component } from 'react';
import { withStyles } from 'material-ui/styles';
import styles from './SpeedDial.styles';
import type { ClassesType } from './SpeedDial.styles';
import type { SpeedDialPropsType, StateEnumType, SpeedDialStateType } from '../../../material-ui-speed-dial.js.flow';

export type SpeedDialWithStylePropsType = {
	classes: ClassesType,
};
export type SpeedDialAllPropsType = SpeedDialWithStylePropsType & SpeedDialPropsType;

export type ForceSetStateType = ({ state: StateEnumType }, cb?: () => void) => void;

const STATE = {
	CLOSED: 'closed',
	OPENING: 'opening',
	OPENED: 'opened',
	CLOSING: 'closing',
};

export class SpeedDial extends Component<SpeedDialAllPropsType, SpeedDialStateType> {
	static displayName = 'SpeedDial';
	static defaultProps = {
		animationDelay: 500,
	};

	state = {
		state: STATE.CLOSED,
	};

	componentDidMount() {
		this.forceSetState = this.setState;
	}

	componentWillUnmount() {
		/* istanbul ignore next */
		this.forceSetState = () => {};
	}

	handleOpen = () => {
		this.forceSetState({ state: STATE.OPENING }, () => {
			setTimeout(() => {
				this.forceSetState({ state: STATE.OPENED });
			}, this.props.animationDelay);
		});
	};
	handleClose = () => {
		this.forceSetState({ state: STATE.CLOSING }, () => {
			setTimeout(() => {
				this.forceSetState({ state: STATE.CLOSED });
			}, this.props.animationDelay);
		});
	};

	forceSetState: ForceSetStateType;

	renderButton(): React$Element<*> | Array<React$Element<*>> {
		const { classes, renderButton, renderOpenedButton } = this.props;
		const { state } = this.state;

		if (!renderOpenedButton) {
			return renderButton(
				{
					className: `${classes.button}`,
					onClick: state === STATE.OPENED ? this.handleClose : this.handleOpen,
				},
				{
					className: `${classes.buttonIcon} ${classes.buttonIconSingle} ${
						classes[`buttonIconSingle--state-${state}`]
					}`,
				}
			);
		}

		return [
			renderButton(
				{
					key: 'closed',
					className: `${classes.button} ${classes.buttonClosed} ${classes[`buttonClosed--state-${state}`]}`,
					onClick: this.handleOpen,
				},
				{
					className: `${classes.buttonIcon} ${classes.buttonClosedIcon} ${
						classes[`buttonClosedIcon--state-${state}`]
					}`,
				}
			),
			renderOpenedButton(
				{
					key: 'open',
					className: `${classes.button} ${classes.buttonOpened} ${classes[`buttonOpened--state-${state}`]}`,
					onClick: this.handleClose,
				},
				{
					className: `${classes.buttonIcon} ${classes.buttonOpenedIcon} ${
						classes[`buttonOpenedIcon--state-${state}`]
					}`,
				}
			),
		];
	}

	render(): React$Element<*> {
		const { classes, className, renderList, children } = this.props;
		const { state } = this.state;
		return (
			<div className={className}>
				{renderList({
					className: `${classes.list} ${classes[`list--state-${state}`]}`,
					children: children({
						state,
						className: '',
					}),
				})}
				{this.renderButton()}
			</div>
		);
	}
}

export default withStyles(styles)(SpeedDial);
