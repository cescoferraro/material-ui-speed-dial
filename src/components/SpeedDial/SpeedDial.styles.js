// @flow
import type { StylesType, ThemeType } from '../../types/styles';

let stylesRaw;

export const styles = (theme: ThemeType): StylesType => {
	stylesRaw = {
		root: {
			position: 'fixed',
			bottom: theme.spacing.unit * 2,
			right: theme.spacing.unit * 2,
		},
		button: {
			position: 'absolute',
		},
	};
	return stylesRaw;
};

export type ClassesType = { [key: $Keys<typeof stylesRaw>]: string };

export default styles;
