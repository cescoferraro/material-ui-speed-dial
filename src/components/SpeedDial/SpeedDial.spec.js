/* eslint react/display-name: 0*/
import React from 'react';
import { shallow } from 'enzyme';
import { SpeedDial } from './SpeedDial';
import styles from './SpeedDial.styles';

const props = {
	children: () => {},
	className: 'className',
	renderButton: propsBtn => <button {...propsBtn} />,
	renderList: () => {},
	classes: styles,
};

describe('SpeedDial', () => {
	it('snapshot: default', () => {
		const _props = { ...props };
		const tree = shallow(<SpeedDial {..._props} />);
		expect(tree).toMatchSnapshot();
		expect(tree.find('button').length).toBe(1);
	});
	it('snapshot: without className', () => {
		const _props = { ...props, className: undefined };
		const tree = shallow(<SpeedDial {..._props} />);
		expect(tree).toMatchSnapshot();
	});
	it('snapshot: renderOpenedButton', () => {
		const _props = { ...props, renderOpenedButton: propsBtn => <button {...propsBtn} /> };
		const tree = shallow(<SpeedDial {..._props} />);
		expect(tree).toMatchSnapshot();
		expect(tree.find('button').length).toBe(2);
	});
	it('simulate: click open', done => {
		const _props = { ...props };
		const tree = shallow(<SpeedDial {..._props} />);
		expect(tree.find('button').length).toBe(1);
		tree.find('button').simulate('click');
		expect(tree.state('state')).toBe('opening');
		setTimeout(() => {
			expect(tree.state('state')).toBe('opened');
			done();
		}, 500);
	});
	it('simulate: click close', done => {
		const _props = { ...props };
		const tree = shallow(<SpeedDial {..._props} />);
		tree.setState({
			state: 'opened',
		});
		tree.find('button').simulate('click');
		expect(tree.state('state')).toBe('closing');
		setTimeout(() => {
			expect(tree.state('state')).toBe('closed');
			done();
		}, 500);
	});
	it('simulate: click open with two buttons', done => {
		const _props = { ...props, renderOpenedButton: propsBtn => <button {...propsBtn} /> };
		const tree = shallow(<SpeedDial {..._props} />);
		expect(tree.find('button').length).toBe(2);
		tree
			.find('button')
			.first()
			.simulate('click');
		expect(tree.state('state')).toBe('opening');
		setTimeout(() => {
			expect(tree.state('state')).toBe('opened');
			done();
		}, 500);
	});
	it('snapshot: renderBackdrop', () => {
		const _props = {
			...props,
			renderBackdrop: propsBackdrop => (
				<div className="backdrop" {...{ ...propsBackdrop, className: `${propsBackdrop.className} backdrop` }} />
			),
		};
		const tree = shallow(<SpeedDial {..._props} />);
		expect(tree).toMatchSnapshot();
		expect(tree.find('.backdrop').length).toBe(1);
	});
	it('simulate: close by backdrop', done => {
		const _props = {
			...props,
			renderBackdrop: propsBackdrop => (
				<div className="backdrop" {...{ ...propsBackdrop, className: `${propsBackdrop.className} backdrop` }} />
			),
		};
		const tree = shallow(<SpeedDial {..._props} />);
		tree.setState({
			state: 'opened',
		});
		tree.find('.backdrop').simulate('click');
		setTimeout(() => {
			expect(tree.state('state')).toBe('closed');
			done();
		}, 500);
	});
	it('snapshot: renderOuterRimButton', () => {
		const _props = {
			...props,
			renderOuterRimButton: propsOuterRimButtom => (
				<div
					className="backdrop"
					{...{ ...propsOuterRimButtom, className: `${propsOuterRimButtom.className} outerRimButtom` }}
				/>
			),
		};
		const tree = shallow(<SpeedDial {..._props} />);
		expect(tree).toMatchSnapshot();
		expect(tree.find('.outerRimButtom').length).toBe(1);
	});
	it('unmount should work without an error while animation', done => {
		const _props = { ...props };
		const tree = shallow(<SpeedDial {..._props} />);
		tree.find('button').simulate('click');
		setTimeout(() => {
			tree.unmount();
			done();
		}, 100);
	});
	it('styles', () => {
		expect(typeof styles()).toBe('object');
	});
});
