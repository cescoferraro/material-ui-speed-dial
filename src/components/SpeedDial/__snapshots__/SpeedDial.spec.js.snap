// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeedDial snapshot: default 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SpeedDial
    animationDelay={500}
    className="className"
    classes={[Function]}
    renderButton={[Function]}
    renderList={[Function]}
>
    [Function]
</SpeedDial>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        <button
          className="undefined"
          onClick={[Function]}
/>,
      ],
      "className": "className",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "undefined",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <button
            className="undefined"
            onClick={[Function]}
/>,
        ],
        "className": "className",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "undefined",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`SpeedDial snapshot: renderOpenedButton 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SpeedDial
    animationDelay={500}
    className="className"
    classes={[Function]}
    renderButton={[Function]}
    renderList={[Function]}
    renderOpenedButton={[Function]}
>
    [Function]
</SpeedDial>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        Array [
          <button
            className="undefined undefined undefined"
            onClick={[Function]}
/>,
          <button
            className="undefined undefined undefined"
            onClick={[Function]}
/>,
        ],
      ],
      "className": "className",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      Object {
        "instance": null,
        "key": "closed",
        "nodeType": "host",
        "props": Object {
          "className": "undefined undefined undefined",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "button",
      },
      Object {
        "instance": null,
        "key": "open",
        "nodeType": "host",
        "props": Object {
          "className": "undefined undefined undefined",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          Array [
            <button
              className="undefined undefined undefined"
              onClick={[Function]}
/>,
            <button
              className="undefined undefined undefined"
              onClick={[Function]}
/>,
          ],
        ],
        "className": "className",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": "closed",
          "nodeType": "host",
          "props": Object {
            "className": "undefined undefined undefined",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
        Object {
          "instance": null,
          "key": "open",
          "nodeType": "host",
          "props": Object {
            "className": "undefined undefined undefined",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
